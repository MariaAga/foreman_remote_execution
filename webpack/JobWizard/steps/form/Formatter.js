import React from 'react';
import PropTypes from 'prop-types';
import { FormGroup, TextInput, TextArea } from '@patternfly/react-core';
import { helpLabel } from './FormHelpers';
import { SelectField } from './SelectField';
import { TemplateSearchField } from './TemplateSearchField';

export const Formatter = ({ input, values, setValue }) => {
  const isSelectType = !!input?.options;
  const inputType = input.value_type;
  const isTextType = inputType === 'plain' || !inputType; // null defaults to plain

  const { name, required, hidden_value: hidden } = input;
  const labelText = input.description;
  const value = values[name];
  const id = name.replace(/ /g, '-');
  if (isSelectType) {
    const options = input.options.split(/\r?\n/).map(option => option.trim());
    return (
      <SelectField
        aria-label={name}
        key={id}
        isRequired={required}
        label={name}
        fieldId={id}
        options={options}
        labelIcon={helpLabel(labelText, name)}
        value={value}
        setValue={newValue => setValue({ ...values, [name]: newValue })}
      />
    );
  }
  if (isTextType) {
    return (
      <FormGroup
        key={name}
        label={name}
        labelIcon={helpLabel(labelText, name)}
        fieldId={id}
        isRequired={required}
      >
        <TextArea
          aria-label={name}
          className={hidden ? 'masked-input' : null}
          required={required}
          rows={2}
          id={id}
          value={value}
          onChange={newValue => setValue({ ...values, [name]: newValue })}
        />
      </FormGroup>
    );
  }
  if (inputType === 'date') {
    return (
      <FormGroup
        key={name}
        label={name}
        labelIcon={helpLabel(labelText, name)}
        fieldId={id}
        isRequired={required}
      >
        <TextInput
          aria-label={name}
          placeholder="YYYY-mm-dd HH:MM"
          className={hidden ? 'masked-input' : null}
          required={required}
          id={id}
          type="text"
          value={value}
          onChange={newValue => setValue({ ...values, [name]: newValue })}
        />
      </FormGroup>
    );
  }
  if (inputType === 'search') {
    const controller = input.resource_type;
    // TODO: get text from redux autocomplete
    return (
      <TemplateSearchField
        key={id}
        name={name}
        defaultValue={value}
        controller={controller}
        labelText={labelText}
        required={required}
        setValue={setValue}
        values={values}
      />
    );
  }

  return null;
};

Formatter.propTypes = {
  input: PropTypes.object.isRequired,
  setValue: PropTypes.func.isRequired,
  values: PropTypes.object.isRequired,
};
