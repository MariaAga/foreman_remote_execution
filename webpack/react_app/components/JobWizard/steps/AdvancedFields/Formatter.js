import React, { useEffect } from 'react';
import { useSelector } from 'react-redux';
import { FormGroup, TextInput } from '@patternfly/react-core';
import SearchBar from 'foremanReact/components/SearchBar';
import { helpLabel } from '../form/FormHelpers';
import { SelectField } from '../form/SelectField';

const TemplateSearchField = ({
  name,
  controller,
  labelText,
  required,
  defaultValue,
  setValue,
  values,
}) => {
  const searchQuery = useSelector(
    state => state.autocomplete[name]?.searchQuery
  );
  useEffect(() => {
    setValue({ ...values, [name]: searchQuery });
  }, [searchQuery]);
  return (
    <FormGroup
      key={name}
      label={name}
      labelIcon={helpLabel(labelText)}
      fieldId={name}
      isRequired={required}
    >
      <SearchBar
        initialQuery={defaultValue}
        data={{
          controller,
          autocomplete: {
            id: name,
            url: `/${controller}/auto_complete_search`,
            useKeyShortcuts: true,
          },
        }}
      />
    </FormGroup>
  );
};

export const formatter = (input, values, setValue) => {
  const isSelectType = !!input?.options;
  const inputType = input.value_type;
  const isTextType = inputType === 'plain' || !inputType; // null defaults to plain

  const { name, required, hidden_value: hidden } = input;
  const labelText = input.description;
  const defaultValue = input.default;
  const value = values[name] || defaultValue;

  if (isSelectType) {
    const options = input.options.split(/\r?\n/).map(option => option.trim());
    return (
      <SelectField
        key={name}
        isRequired={required}
        label={name}
        fieldId={name}
        options={options}
        labelText={labelText}
        value={value}
        setValue={newValue => setValue({ ...values, [name]: newValue })}
      />
    );
  }
  if (isTextType) {
    return (
      <FormGroup
        key={name}
        label={name}
        labelIcon={helpLabel(labelText)}
        fieldId={name}
        isRequired={required}
      >
        <TextInput
          className={hidden ? 'masked-input' : null}
          required={required}
          rows={2}
          id={name}
          type="text"
          value={value}
          onChange={newValue => setValue({ ...values, [name]: newValue })}
        />
      </FormGroup>
    );
  }
  if (inputType === 'date') {
    return <span key={name}>DateTime here</span>;
  }
  if (inputType === 'search') {
    const controller = input.resource_type;
    // TODO: get text from redux autocomplete
    return (
      <TemplateSearchField
        name={name}
        controller={controller}
        labelText={labelText}
        required={required}
        defaultValue={defaultValue}
        setValue={setValue}
        values={values}
      />
    );
  }

  return null;
};
