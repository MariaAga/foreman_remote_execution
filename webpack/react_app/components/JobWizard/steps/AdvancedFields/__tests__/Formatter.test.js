import { testComponentSnapshotsWithFixtures } from '@theforeman/test';
import { formatter } from '../Formatter';

const baseFixture = {
  id: 11,
  name: 'template field',
};

const fixtures = {
  text: {
    ...baseFixture,
    value_type: 'plain',
    default: 'some default',
    hidden_value: true,
  },
  select: {
    ...baseFixture,
    value_type: 'plain',
    advanced: true,
    options: 'a\r\nb',
  },
  search: {
    ...baseFixture,
    value_type: 'search',
    resource_type: 'ansible_roles',
  },
  datetime: { ...baseFixture, value_type: 'date', required: true },
};

describe('formatter', () =>
  testComponentSnapshotsWithFixtures(formatter, fixtures));
