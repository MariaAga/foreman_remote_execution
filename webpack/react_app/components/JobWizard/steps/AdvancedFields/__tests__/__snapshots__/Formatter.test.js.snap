// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatter datetime 1`] = `
<span>
  DateTime here
</span>
`;

exports[`formatter search 1`] = `
<Fragment>
  <FormGroup
    fieldId="template field"
    label="template field"
    labelIcon={null}
  >
    <SearchBar
      data={
        Object {
          "autocomplete": Object {
            "id": "template field",
            "url": "/ansible_roles/auto_complete_search",
            "useKeyShortcuts": true,
          },
          "controller": "ansible_roles",
        }
      }
      key="template field"
    />
  </FormGroup>
</Fragment>
`;

exports[`formatter select 1`] = `
<SelectField
  fieldId="template field"
  key="template field"
  label="template field"
  options={
    Array [
      "a",
      "b",
    ]
  }
  setValue={[Function]}
  value={null}
/>
`;

exports[`formatter text 1`] = `
<FormGroup
  fieldId="template field"
  label="template field"
  labelIcon={null}
>
  <TextInput
    className="masked-input"
    id="template field"
    key="template field"
    onChange={[Function]}
    rows={2}
    type="text"
    value="some default"
  />
</FormGroup>
`;
